feat: add payment dashboard

tnrdd opened the pull request:
# Payment Dashboard

## Description

Add a dashboard where the users can see and manage their parcels.

fixes #104, fixes #235 

## Checklist:

- [x] My commit message follows the Conventional Commits specificaiton
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] My PR is rebased off the most recent `feat/acl-architecture` branch
- [x] My PR is opened against the `feat/acl-architecture` branch

## Alert Reviewers

@codynhat @gravenp

gravenp commented:
Thanks, @tnrdd!

This is looking good, but here are few potential open items:

- It doesn't look like `Incoming Bids` that are rejected are being handled correctly (i.e. instead of returning to `Valid` for the licensor the parcels still show the `Incoming Bid` info in the table and show an entry for `Outgoing Bid` for the bidder). When a bid is accepted, the parcel is correctly removed & shown for the previous owner and new owner respectively
- I'm not seeing any notification badges. This includes when there are parcels that require action besides the "incorrectly classified" parcels in scenario described above.
- Is it possible to zoom to the parcel from the Dashboard button with the parcel a bit more centered? Right now it appears that the bottom left (SW corner) of the parcel is placed in the center-ish of the map view. Calculating the mid-point of the parcel seems overly complex, but if we always centered on a few coordinates north and east of the bottom left it'd look a bit better.
- I need to wait for a bid to age 7 days to check out the `Needs Transfer` status and am working through foreclosure scenarios.

tnrdd commented:
Thanks for the review @gravenp!
> * It doesn't look like `Incoming Bids` that are rejected are being handled correctly (i.e. instead of returning to `Valid` for the licensor the parcels still show the `Incoming Bid` info in the table and show an entry for `Outgoing Bid` for the bidder). When a bid is accepted, the parcel is correctly removed & shown for the previous owner and new owner respectively
> * I'm not seeing any notification badges. This includes when there are parcels that require action besides the "incorrectly classified" parcels in scenario described above.
>
Both should be fixed now.
> * Is it possible to zoom to the parcel from the Dashboard button with the parcel a bit more centered? Right now it appears that the bottom left (SW corner) of the parcel is placed in the center-ish of the map view. Calculating the mid-point of the parcel seems overly complex, but if we always centered on a few coordinates north and east of the bottom left it'd look a bit better.
> 
I picked the first coordinates we got because the squares that compose a parcel doesn't seem to come in order, otherwise with two corner poins would be easy to calculate the center of the parcel.

However having the center of the parcel coordinates wouldn't have been enough to have the view nicely centered because the sidebar is shifting it.
I have added an offset to the coordinates to compensate for the shift and should be centered now.

gravenp commented:
@tnrdd 

1) Parcels with rejected incoming bids are displaying correctly now. The rejected outgoing bids are still (incorrectly) showing on the bidder's side though. They should be cleared upon rejection.
2) Notification badges look good.
3) The approach you took for centering on zoom seems reasonable and looks better. We can leave this for now and if/when we revisit how parcel paths are rendered/constructed we can look at other options.

gravenp commented:
Couple things I came across when using a fresh address (Github isn't allowing me to upload screenshots again):

1) When an address doesn't own any parcels or have bids outstanding, change the text that displays where the table would be to "No parcels or bids yet..."
2) When an address doesn't have any parcels and therefore any active streams, don't display the ETHx Balance warning at the top of the page. Currently, it shows the "0 balance" text presumably because the date calc is undefined.
3) If an address doesn't own a parcel, but _does_ have an outgoing bid, the bid is not showing in the table. Once a parcel is claimed, both the parcel and the claim are showing.

tnrdd commented:
>* Parcels with rejected incoming bids are displaying correctly now. The rejected outgoing bids are still (incorrectly) showing on the bidder's side though. They should be cleared upon rejection.
>* When an address doesn't own any parcels or have bids outstanding, change the text that displays where the table would be to "No parcels or bids yet..."
>* When an address doesn't have any parcels and therefore any active streams, don't display the ETHx Balance warning at the top of the page. Currently, it shows the "0 balance" text presumably because the date calc is undefined.
> 
Fixed the outgoing bid bug and changed the info messages.
> If an address doesn't own a parcel, but does have an outgoing bid, the bid is not showing in the table. Once a parcel is claimed, both the parcel and the claim are showing.
>
I wasn't able to reproduce this, it might have been a bit of lag and the update was missed.
I made a change on the update logic so that when the portfolio needs to be updated we refetch every few seconds until there is new data.

gravenp commented:
These last changes close everything out. Great work @tnrdd!

@codynhat, this PR is good from the user side if you can review and merge when appropriate.

codynhat commented:
@tnrdd I just merged the `feat/acl-architecture` branch into `develop`. Could you rebase this against `develop` to resolve the conflicts?

codynhat commented on components/profile/ProfileModal.tsx:
```diff
@@ -1,105 +1,866 @@
+          if (licenseOwner === account && !isPayerBidActive) {
+            status = "In Foreclosure";
+            actionDate = dayjs.unix(currentBid.timestamp).format("YYYY/MM/D");
+            forSalePrice = calculateAuctionValue(
```
The logic for calculating the current foreclosure auction price has been updated https://github.com/Geo-Web-Project/cadastre/blob/develop/components/cards/AuctionInfo.tsx#L80

This may be updated again soon, so we should probably put it in a shared place.

tnrdd replied:
I put `calculateAuctionValue()` in `/lib/utils.ts`, is that ok?

codynhat replied:
That is fine, but see the change in logic. The auction start time and length are different now.

tnrdd replied:
I see, should I write a function that returns an object with start time and auction length to be reused or just use the same logic here?

codynhat replied:
You could keep `calculateAuctionValue` as is I suppose, but just update this to pass in the correct start time and length. But the logic for getting the start time and length should be defined in a single place.

codynhat reviewed (changes requested):
I have a few parcels in foreclosure

tnrdd commented on components/profile/ProfileModal.tsx:
```diff
@@ -1,105 +1,866 @@
+            const deadline = Number(pendingBid.timestamp) + SECONDS_IN_WEEK;
+            const isPastDeadline = deadline * 1000 <= Date.now();
+            if (isPastDeadline || ownerBidContributionRate.eq(0)) {
+              status = "Needs Tranfer";
```
Fixed it, thanks.

tnrdd commented on components/profile/ProfileModal.tsx:
```diff
@@ -1,105 +1,866 @@
+                  Date.now() / 1000
+              ? `At the current rate, your ETHx balance will reach 0 on ${dayjs
+                  .unix(accountTokenSnapshot.maybeCriticalAtTimestamp)
+                  .format("MMM D, YYYY h:mmA.")}`
```
After #239 is closed it will be displayed using the user's local time, should it be specified a timezone anyway?

codynhat replied:
Yes. The user needs to know it's in their local timezone and not some other timezone

tnrdd commented on components/profile/ProfileModal.tsx:
```diff
@@ -1,105 +1,866 @@
+                    ? deadline
+                    : currentBid.timestamp
+                )
+                .format("YYYY/MM/D");
```
I don't have a preference, I'm more used to slashes but I think in other places it is more common with the dashes.
In `AuctionInfo.tsx` we use slashes so for consistency we may want to change it there too if we want to go with dashes.

gravenp replied:
We can use dashes if that's more common. I don't have a preference, but agree let's just be consistent.

codynhat replied:
Either way, let's add `DD` so it contains the 0

codynhat commented on pages/index.tsx:
```diff
@@ -162,11 +175,26 @@ function IndexPage() {
-        />
+        <>
+          {ceramic && registryContract && paymentToken && library && (
+            <Profile
```
@tnrdd Looks like this one still needs to be updated

tnrdd replied:
Can it be done on the index page too? All the properties are defined here and passed individually to the map too.

codynhat replied:
Oh sorry, never mind

codynhat commented:
Great work @tnrdd!
